crates/nargo_cli/src/cli/fs/common_reference_string.rs

```diff
@@ -0,0 +1,55 @@
+use std::{env, path::PathBuf};
+
+use acvm::{acir::circuit::Circuit, CommonReferenceString};
+
+use super::{create_named_dir, write_to_file};
+
+// TODO: pull this from backend.
+const BACKEND_IDENTIFIER: &str = "acvm-backend-barretenberg";
+const TRANSCRIPT_NAME: &str = "common-reference-string.bin";
+
+fn common_reference_string_location() -> PathBuf {
+    let cache_dir = match env::var("NARGO_BACKEND_CACHE_DIR") {
+        Ok(cache_dir) => PathBuf::from(cache_dir),
+        Err(_) => dirs::home_dir().unwrap().join(".nargo").join("backends"),
+    };
+    cache_dir.join(BACKEND_IDENTIFIER).join(TRANSCRIPT_NAME)
+}
+
+pub(crate) fn read_cached_common_reference_string() -> Vec<u8> {
+    let crs_path = common_reference_string_location();
+
+    // TODO: Implement checksum
+    match std::fs::read(crs_path) {
+        Ok(common_reference_string) => common_reference_string,
+        Err(_) => vec![],
+    }
+}
+
+pub(crate) fn update_common_reference_string<B: CommonReferenceString>(
+    backend: &B,
+    common_reference_string: &[u8],
```

Comment:
```suggestion
    common_reference_string: Vec<u8>,
```

We should pass ownership here similarly to how we do to the backend, in either outcome we shouldn't use the original CRS anymore.

Reply:
Looks like this causes issues due to moving `common_references_string` into the closures. Perhaps we should  separate the CRS downloads from the preprocessing for each function?

Reply:
None of these changes would affect external behaviour though so I'm happy to just open an issue and revisit later as getting ACVM merged is more important.